File: tests_deps/src/main/java/com/github/klee0kai/test/car/di/cachecontrol/gc/WheelMultiMappedGcModule.java
package com.github.klee0kai.test.car.di.cachecontrol.gc;

import com.github.klee0kai.stone.annotations.module.Module;
import com.github.klee0kai.stone.annotations.module.Provide;
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWheelScope;
import com.github.klee0kai.test.car.model.Wheel;

import java.lang.ref.WeakReference;

@Module
public class WheelMultiMappedGcModule {

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Factory)
    public WeakReference<Wheel> wheelFactory(String q1, Integer q2) {
        return new WeakReference<>(new Wheel());
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Weak)
    public WeakReference<Wheel> wheelWeak(String q1, Integer q2) {
        return new WeakReference<>(new Wheel());
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Soft)
    public WeakReference<Wheel> wheelSoft(String q1, Integer q2) {
        return new WeakReference<>(new Wheel());
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Strong)
    public WeakReference<Wheel> wheelStrong(String q1, Integer q2) {
        return new WeakReference<>(new Wheel());
    }


}

File: tests_wraps_kotlin/src/main/kotlin/com/github/klee0kai/test/car/di/cachecontrol/gc/WheelMultiMappedGcModule.kt
package com.github.klee0kai.test.car.di.cachecontrol.gc

import com.github.klee0kai.stone.annotations.module.Module
import com.github.klee0kai.stone.annotations.module.Provide
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWheelScope
import com.github.klee0kai.test.car.model.Wheel
import java.lang.ref.WeakReference

@Module
open class WheelMultiMappedGcModule {
    @GcWheelScope
    @Provide(cache = Provide.CacheType.Factory)
    open fun wheelFactory(q1: String?, q2: Integer?): WeakReference<Wheel> {
        return WeakReference(Wheel())
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Weak)
    open fun wheelWeak(q1: String, q2: Integer): WeakReference<Wheel> {
        return WeakReference(Wheel())
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Soft)
    open fun wheelSoft(q1: String, q2: Integer): WeakReference<Wheel> {
        return WeakReference(Wheel())
    }

    @GcWheelScope
    @Provide(cache = Provide.CacheType.Strong)
    open fun wheelStrong(q1: String, q2: Integer?): WeakReference<Wheel> {
        return WeakReference(Wheel())
    }
}
